modules/phase_field/src/action/GrandPotentialKernelAction.C

```diff
@@ -244,50 +262,136 @@ GrandPotentialKernelAction::act()
+        _problem->addKernel("TimeDerivative", kernel_name, params);
+     
+        // MatDiffusion concentration (coupled with chempot)
+        params = _factory.getValidParams("MatDiffusion");
+        params.set<NonlinearVariableName>("variable") = c_names[i];
+        params.set<std::vector<VariableName>>("v") = v3;
+        params.set<bool>("implicit") = implicity;
+        params.set<bool>("use_displaced_mesh") = displaced_mesh;
+        params.set<MaterialPropertyName>("diffusivity") = M[i];
+        kernel_name = "MatDif_" + w_names[i];
+        if (aniso.get(i))
+        {
+          params.set<std::vector<VariableName>>("args") = v1;
+          _problem->addKernel("MatAnisoDiffusion", kernel_name, params);
+        }
+        else
+        {
+           params.set<std::vector<VariableName>>("args") = v0;
+           _problem->addKernel("MatDiffusion", kernel_name, params);
+        }
+      } // for (unsigned int i = 0; i < n_c; ++i)
+      
+      // Chemical Potentials
+      for (unsigned int i = 0; i < n_w; ++i)
+      {
+        // coupling of c and w
+        InputParameters params = _factory.getValidParams("MatReaction");
+        params.set<NonlinearVariableName>("variable") = w_names[i];
+        params.set<std::vector<VariableName>>("v") = v4;
+        params.set<MaterialPropertyName>("mob_name") = mob_cw;
+        kernel_name = "MR_c" + w_names[i];
+        _problem->addKernel("MatReaction", kernel_name, params);
+         
+        // contribution between chempot and each grains to concentration
+        // Summations of MatReaction and MaskedBodyForce
+         
+        for (unsigned int j = 0; j < n_hj; ++j)
+        {
+          // MatReaction
+          params = _factory.getValidParams("MatReaction");
+          params.set<NonlinearVariableName>("variable") = w_names[i];
+          params.set<std::vector<VariableName>>("args") = v1;
+          params.set<MaterialPropertyName>("mob_name") = MRj[j];
+          params.set<std::vector<VariableName>>("args") = v1;
+          params.set<bool>("implicit") = implicity;
+          params.set<bool>("use_displaced_mesh") = displaced_mesh;
+          kernel_name = "MR_" + w_names[i] + "_" + all_etas[j];
+          _problem->addKernel("MatReaction", kernel_name, params);
+          
+          //MaskedBodyForce
+          params = _factory.getValidParams("MaskedBodyForce");
+          params.set<NonlinearVariableName>("variable") = w_names[i];
+          params.set<std::vector<VariableName>>("args") = v1;
+          params.set<MaterialPropertyName>("mask") = BFj[j];
+          params.set<bool>("implicit") = implicity;
+          params.set<bool>("use_displaced_mesh") = displaced_mesh;
+          kernel_name = "MBD_" + w_names[i] + "_" + all_etas[j];
+          _problem->addKernel("MaskedBodyForce", kernel_name, params);
+        }
+      } // for (unsigned int i = 0; i < n_w; ++i)
+    } // case true
+  } // switch
 } // GrandPotentialKernelAction::act()
```

Comment:
Please remove all these end-block comments. If the code is not readable enough  with just the correct indentation then the blocks span too many lines and the code should be refactored.